esphome/components/sx127x/__init__.py

```diff
@@ -0,0 +1,332 @@
+            cv.Optional(CONF_DIO0_PIN): pins.internal_gpio_input_pin_schema,
+            cv.Required(CONF_FREQUENCY): cv.int_range(min=137000000, max=1020000000),
+            cv.Required(CONF_MODULATION): cv.enum(MOD),
+            cv.Optional(CONF_ON_PACKET): automation.validate_automation(single=True),
+            cv.Optional(CONF_PA_PIN, default="BOOST"): cv.enum(PA_PIN),
+            cv.Optional(CONF_PA_POWER, default=17): cv.int_range(min=0, max=17),
+            cv.Optional(CONF_PA_RAMP, default="40us"): cv.enum(RAMP),
+            cv.Optional(CONF_PAYLOAD_LENGTH, default=0): cv.int_range(min=0, max=256),
+            cv.Optional(CONF_PREAMBLE_DETECT, default=0): cv.int_range(min=0, max=3),
+            cv.Optional(CONF_PREAMBLE_ERRORS, default=0): cv.int_range(min=0, max=31),
+            cv.Optional(CONF_PREAMBLE_POLARITY, default=0xAA): cv.All(
+                cv.hex_int, cv.one_of(0xAA, 0x55)
+            ),
+            cv.Optional(CONF_PREAMBLE_SIZE, default=0): cv.int_range(min=0, max=65535),
+            cv.Required(CONF_RST_PIN): pins.internal_gpio_output_pin_schema,
+            cv.Optional(CONF_RX_FLOOR, default=-94): cv.float_range(min=-128, max=-1),
+            cv.Optional(CONF_RX_START, default=True): cv.boolean,
+            cv.Optional(CONF_SHAPING, default="NONE"): cv.enum(SHAPING),
+            cv.Optional(CONF_SPREADING_FACTOR, default=7): cv.int_range(min=6, max=12),
+            cv.Optional(CONF_SYNC_VALUE, default=[]): cv.ensure_list(cv.hex_uint8_t),
+        },
+    )
+    .extend(cv.COMPONENT_SCHEMA)
+    .extend(spi.spi_device_schema(True, 8e6, "mode0"))
+    .add_extra(validate_config)
+)
+
+
+async def to_code(config):
+    var = cg.new_Pvariable(config[CONF_ID])
+    await cg.register_component(var, config)
+    await spi.register_spi_device(var, config)
+    if CONF_ON_PACKET in config:
+        await automation.build_automation(
+            var.get_packet_trigger(),
+            [
+                (cg.std_vector.template(cg.uint8), "x"),
+                (cg.float_, "rssi"),
+                (cg.float_, "snr"),
+            ],
+            config[CONF_ON_PACKET],
+        )
+    if CONF_DIO0_PIN in config:
+        dio0_pin = await cg.gpio_pin_expression(config[CONF_DIO0_PIN])
+        cg.add(var.set_dio0_pin(dio0_pin))
+    rst_pin = await cg.gpio_pin_expression(config[CONF_RST_PIN])
+    cg.add(var.set_rst_pin(rst_pin))
+    cg.add(var.set_auto_cal(config[CONF_AUTO_CAL]))
+    cg.add(var.set_bandwidth(config[CONF_BANDWIDTH]))
+    cg.add(var.set_frequency(config[CONF_FREQUENCY]))
+    cg.add(var.set_deviation(config[CONF_DEVIATION]))
+    cg.add(var.set_modulation(config[CONF_MODULATION]))
+    cg.add(var.set_pa_pin(config[CONF_PA_PIN]))
+    cg.add(var.set_pa_ramp(config[CONF_PA_RAMP]))
+    cg.add(var.set_pa_power(config[CONF_PA_POWER]))
+    cg.add(var.set_shaping(config[CONF_SHAPING]))
+    if CONF_BITRATE in config:
+        cg.add(var.set_bitrate(config[CONF_BITRATE]))
```

Comment:
Just a suggestion, makes the code a bit cleaner (in my oppinion).
```suggestion
    if bitrate := config.get(CONF_BITRATE):
        cg.add(var.set_bitrate(bitrate))
```

Reply:
It looks more consistent with the other lines though:
```
    cg.add(var.set_pa_ramp(config[CONF_PA_RAMP]))
    cg.add(var.set_pa_power(config[CONF_PA_POWER]))
    cg.add(var.set_shaping(config[CONF_SHAPING]))
    if CONF_BITRATE in config:
        cg.add(var.set_bitrate(config[CONF_BITRATE]))
    else:
        cg.add(var.set_bitrate(4800))
```

Reply:
You should be setting the default inside the config_schema, then you can assume there is always a value. Same for the below ones too

Reply:
I did this because I made a breaking change and needed to know if it was set in the validation code. It's been long enough though. I will clean it up.